README.md
# RVL-Node Animations

A set of helper methods to create and render animations for [RVL-Node](https://github.com/nebrius/RVL-Node) and friends

## System Requirements

- Any modern JavaScript engine (web or Node.js)

## Installation

Install using npm:

```bash
npm install rvl-node-animations
```

## Usage

Basic usage:

```typescript
import {
  initRenderer,
  renderPixels,
  createWaveParameters,
  createSolidColorWave,
  createColorCycleWave
} from 'rvl-node-animations';

// Initialize the renderer
initRenderer();

const waveParameters = createWaveParameters(
  // Create a solid-cyan, half-way transparent color
  createSolidColorWave(128, 255, 128),

  // Create a fully opaque, slow color cycle that will show throw the cyan wave
  createColorCycleWave(2, 255)
);

// Render the pixels at 33 frames per second
setInterval(() => {
  const pixels = renderPixels(waveParameters);
  // Display pixels in the browser/an LED strip/etc.
}, 33);

```

If you intend to create these animations on the same device that will be controlling the LEDs, here is an example integrating rvl-node-animations and [rvl-node](https://github.com/nebrius/rvl-node):

```typescript
import {
  createWaveParameters,
  createSolidColorWave,
  createColorCycleWave
} from 'rvl-node-animations';
import { RVL } from 'rvl-node';

const rvl = new RVL({
  networkInterface: 'wlan0',
  logLevel: 'debug',
  mode: 'controller'
});

rvl.on('initialized', () => {
  rvl.setWaveParameters(createWaveParameters(
    // Create a solid-cyan, half-way transparent color
    createSolidColorWave(128, 255, 128),

    // Create a fully opaque, slow color cycle that will show throw the cyan wave
    createColorCycleWave(2, 255)
  ));
});
```

## Notes

If you're using TypeScript, you don't need to install separate `@types` type definitions because this package ships with them. To support sharing of interfaces between this module and `rvl-node` itself, I created another module called [rvl-node-types](https://github.com/nebrius/rvl-node-types), which may be of interest to you.

All waves are defined in the [HSV color space](https://en.wikipedia.org/wiki/HSL_and_HSV) that has been mapped to 8-bit integers (e.g. the hue is 0-255, not 0-360, and saturation/value are 0-255, not 0-1 or 0-100). Please familiarize yourself with the HSV color space before using this library.

## API

Note: API signatures are declared using [TypeScript syntax](https://www.typescriptlang.org/). This provides a concise, formal definition for all signatures in a syntax that is familiar to at least some developers.

### createWaveParameters(wave1, wave2, wave3, wave4)

This method creates a fully formed set of wave parameters that can be passed directly to `rvl-node#setWaveParameters()`.

_Signature:_

```typescript
function createWaveParameters(
  wave1?: IWave,
  wave2?: IWave,
  wave3?: IWave,
  wave4?: IWave
): IWaveParameters
```

_Arguments:_

<table>
  <thead>
    <tr>
      <th>Argument</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>wave1 (optional)</td>
      <td>IWave</td>
      <td>The wave parameters for the top-most wave in the stack. If this wave is fully opaque, it will obscure the other waves. Use one of the <pre>create*Wave</pre> methods to generate it. If this parameter is not supplied, a fully transparent wave is created instead.</td>
    </tr>
    <tr>
      <td>wave2 (optional)</td>
      <td>IWave</td>
      <td>The wave parameters for the second top-most wave in the stack. If this parameter is not supplied, a fully transparent wave is created instead.</td>
    </tr>
    <tr>
      <td>wave3 (optional)</td>
      <td>IWave</td>
      <td>The wave parameters for the third top-most wave in the stack. If this parameter is not supplied, a fully transparent wave is created instead.</td>
    </tr>
    <tr>
      <td>wave4 (optional)</td>
      <td>IWave</td>
      <td>The wave parameters for the bottom-most wave in the stack. If this parameter is not supplied, a fully transparent wave is created instead.</td>
    </tr>
  </tbody>
</table>

_Returns:_ an `IWaveParameters` instance. Details can be found at [rvl-node-types](https://github.com/nebrius/rvl-node-types), but for all intents and purposes it can be treated as a black box.

### createEmptyWave()

This method creates an empty wave, aka a fully-transparent wave that allows waves below it in the stack to be fully visible.

_Signature:_

```typescript
function createEmptyWave(): IWave
```

_Arguments:_ none

_Returns:_ an `IWave` instance. Details can be found at [rvl-node-types](https://github.com/nebrius/rvl-node-types), but for all intents and purposes it can be treated as a black box.

### createSolidColorWave(h, s, a)

This method creates a solid color wave that does not change over distance or time.

_Signature:_

```typescript
function createSolidColorWave(h: number, s: number, a: number): IWave
```

_Arguments:_

<table>
  <thead>
    <tr>
      <th>Argument</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>h</td>
      <td>number</td>
      <td>The hue, mapped to between 0 (0 degrees) and 255 (360 degrees).</td>
    </tr>
    <tr>
      <td>s</td>
      <td>number</td>
      <td>The saturation, mapped to between 0 (0%) and 255 (100%).</td>
    </tr>
    <tr>
      <td>a</td>
      <td>number</td>
      <td>The alpha, mapped to between 0 (0%) and 255 (100%).</td>
    </tr>
  </tbody>
</table>

_Returns:_ an `IWave` instance. Details can be found at [rvl-node-types](https://github.com/nebrius/rvl-node-types), but for all intents and purposes it can be treated as a black box.

### createColorCycleWave(rate, a)

This method creates a wave that cycles through colors over time.

From a technical perspective, it linearly varies the hue over time while keeping the saturation and value at 100%.

_Signature:_

```typescript
function createColorCycleWave(rate: number, a: number): IWave
```

_Arguments:_

<table>
  <thead>
    <tr>
      <th>Argument</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>rate</td>
      <td>number</td>
      <td>The rate that the color changes, between 1 and 32. The higher the number, the faster the color changes</td>
    </tr>
    <tr>
      <td>a</td>
      <td>number</td>
      <td>The alpha, mapped to between 0 (0%) and 255 (100%).</td>
    </tr>
  </tbody>
</table>

_Returns:_ an `IWave` instance. Details can be found at [rvl-node-types](https://github.com/nebrius/rvl-node-types), but for all intents and purposes it can be treated as a black box.

### createMovingWave(h, s, rate, spacing)

This method creates a moving wave, with the alpha varying between 0% and 100% over distance, creating a "roving wave" effect.

From a technical perspective, it varies the alpha over distance, and varies the y-value of the start of wave with time.

_Signature:_

```typescript
function createMovingWave(h: number, s: number, rate: number, spacing: number): IWave
```

_Arguments:_

<table>
  <thead>
    <tr>
      <th>Argument</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>h</td>
      <td>number</td>
      <td>The hue, mapped to between 0 (0 degrees) and 255 (360 degrees).</td>
    </tr>
    <tr>
      <td>s</td>
      <td>number</td>
      <td>The saturation, mapped to between 0 (0%) and 255 (100%).</td>
    </tr>
    <tr>
      <td>rate</td>
      <td>number</td>
      <td>The rate that the wave "moves", between 0 and 32. The higher the number, the faster the wave moves. Setting the rate to 0 means the wave does not move at all.</td>
    </tr>
    <tr>
      <td>spacing</td>
      <td>number</td>
      <td>The distance between "peaks" of the waves, between 1 and 16. The higher the number, the shorter the distance.</td>
    </tr>
  </tbody>
</table>

_Returns:_ an `IWave` instance. Details can be found at [rvl-node-types](https://github.com/nebrius/rvl-node-types), but for all intents and purposes it can be treated as a black box.

### createPulsingWave(h, s, rate)

This method creates a wave that pulses, aka it varies from fully transparent to fully opaque over time.

From a technical perspective, it varies the alpha over time with nothing else that varies.

_Signature:_

```typescript
function createPulsingWave(h: number, s: number, rate: number): IWave
```

_Arguments:_

<table>
  <thead>
    <tr>
      <th>Argument</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
  <tbody>
    <tr>
      <td>h</td>
      <td>number</td>
      <td>The hue, mapped to between 0 (0 degrees) and 255 (360 degrees).</td>
    </tr>
    <tr>
      <td>s</td>
      <td>number</td>
      <td>The saturation, mapped to between 0 (0%) and 255 (100%).</td>
    </tr>
    <tr>
      <td>rate</td>
      <td>number</td>
      <td>The rate that the wave "pulses", between 1 and 32. The higher the number, the faster the wave pulses.</td>
    </tr>
  </tbody>
</table>

_Returns:_ an `IWave` instance. Details can be found at [rvl-node-types](https://github.com/nebrius/rvl-node-types), but for all intents and purposes it can be treated as a black box.

### createRainbowWave(a, rate)

This method creates a wave that looks like a rainbow, and slowly moves.

From a technical perspective it varies the hue over distance, and varies the y-value of the start of wave with time.

_Signature:_

```typescript
function createRainbowWave(a: number, rate: number): IWave
```

_Arguments:_

<table>
  <thead>
    <tr>
      <th>Argument</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
  <tbody>
    <tr>
      <td>a</td>
      <td>number</td>
      <td>The alpha, mapped to between 0 (0%) and 255 (100%).</td>
    </tr>
    <tr>
      <td>rate</td>
      <td>number</td>
      <td>The rate that the wave "pulses", between 1 and 32. The higher the number, the faster the wave pulses.</td>
    </tr>
  </tbody>
</table>

_Returns:_ an `IWave` instance. Details can be found at [rvl-node-types](https://github.com/nebrius/rvl-node-types), but for all intents and purposes it can be treated as a black box.

### initRenderer(waveParameters, numPixels, numWaves)

This method initializes the renderer and starts the animation clock.

_Signature:_

```typescript
function initRenderer(waveParameters: IWaveParameters, numPixels: number, numWaves: number = 4): void
```

_Arguments:_

<table>
  <thead>
    <tr>
      <th>Argument</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
  <tbody>
    <tr>
      <td>waveParameters</td>
      <td>number</td>
      <td>An `IWaveParameters` instance describing your animation. Details can be found at [rvl-node-types](https://github.com/nebrius/rvl-node-types), but for all intents and purposes it can be treated as a black box.</td>
    </tr>
    <tr>
      <td>numPixels</td>
      <td>number</td>
      <td>The number of pixels to render. For example, if you are sending this to an LED strip, the number of pixels is the number of LEDs in your strip.</td>
    </tr>
    <tr>
      <td>numWaves (optional)</td>
      <td>number</td>
      <td>The number of wave layers. Default is 4.</td>
    </tr>
  </tbody>
</table>

_Returns:_ none.

### renderPixels()

Renders the pixels for this exact moment in time based on the parameters set in `initRenderer`.

_Signature:_

```typescript
interface IPixel {
  r: number;
  g: number;
  b: number;
}
function renderPixels(): IPixel[]
```

_Arguments:_ none.

_Returns:_ An array of `IPixels` representing the color set in 8-bit RGB format.

### resetRendererClock()

This method resets the renderer clock to 0.

_Signature:_

```typescript
function resetRendererClock(): void
```

_Arguments:_ none.

_Returns:_ none.

### getRendererClock()

Gets the current renderer clock, in milliseconds. This number is not relative to anything, and cannot be used to, e.g., get the wall time.

_Signature:_

```typescript
function getRendererClock(): number
```

_Arguments:_ none.

_Returns:_ The renderer clock as a `number`.

## Background

Note: you do not need to know this information to use this library at all. Some might find it interesting though.

RVL Node uses a rendering engine based on sin waves. RVL Node layers four waves on top of each other to create interesting and aesthetically pleasing LED animations. Think of this like CSS layers. They can include transparency to create a layering effect.

All waves are defined in the HSV color space that has been mapped to 8-bit integers. A hue of 180 degrees is represented by the value `128`, a saturation of 25% is represented by the value `64`, and so on and so forth.

A wave is defined using the following mathematical formula:

```
ledChannelValue(t, x) = a * sin(w_t * t + w_x * x + phi) + b
```

This function takes 5 variables from the user (a, w_x, w_t, phi, b), and 2 from the engine (x, t). This allows someone to create a wave that can vary over time, over the length of the LED strip, can be constant, or any of the above. Each channel (hue, saturation, value, and alpha) have their own wave assigned to them. 4 channels per layer, times 4 layers, means 80 coefficients total!

## License

Copyright (c) Bryan Hughes <bryan@nebri.us>

This file is part of Raver Lights Node Animations.

Raver Lights Node Animations is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

Raver Lights Node Animations is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with Raver Lights Node Animations.  If not, see <http://www.gnu.org/licenses/>.


CHANGELOG.md
## 3.1.1 (2019-8-13)

- Added rendering methods and removed the `clone-deep` dependency.

## 3.0.0 (2019-7-9)

- Welp, that was a mistake. Turns out there were some edge cases in v2.0.0 I missed with brightness that caused flickering animations. I can't fix it with the current approach, so I'm taking brightness out for the time being
- BREAKING CHANGE: removed the `brightness` parameter from all methods.

## 2.0.0 (2019-7-9)

- BREAKING CHANGE: Modified each method so that it's first argument is `brightness`, between 0 and 255
    - The `createSolidColorWave` method no longer takes a `value` argument because it was replaced with the `brightness` argument (which does exactly the same thing under the hood)
- Added the `createRainbowWave` method

## 1.1.0 (2019-5-29)

- Relaxed the `rate` parameter for createMovingWave so that you can now pass in `0` to create a perfectly still wave.

## 1.0.1 (2019-5-16)

- Bug fixes with some animations not rendering correctly

## 1.0.0 (2019-5-16)

- Initial release


declarations/index.d.ts
import { IWaveParameters, IWave } from '../node_modules/rvl-node-types/declarations/index';
export interface IPixel {
    r: number;
    g: number;
    b: number;
}
export declare function initRenderer(waveParameters: IWaveParameters, numPixels: number, numWaves?: number): void;
export declare function resetRendererClock(): void;
export declare function getRendererClock(): number;
export declare function renderPixels(): IPixel[];
export declare function createEmptyWave(): IWave;
export declare function createWaveParameters(wave1?: IWave, wave2?: IWave, wave3?: IWave, wave4?: IWave): IWaveParameters;
export declare function createSolidColorWave(h: number, s: number, a: number): IWave;
export declare function createColorCycleWave(rate: number, a: number): IWave;
export declare function createMovingWave(h: number, s: number, rate: number, spacing: number): IWave;
export declare function createPulsingWave(h: number, s: number, rate: number): IWave;
export declare function createRainbowWave(a: number, rate: number): IWave;


src/index.ts
/*
Copyright (c) Bryan Hughes <bryan@nebri.us>

This file is part of Raver Lights Node Animations.

Raver Lights Node Animations is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

Raver Lights Node Animations is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with Raver Lights Node Animations.  If not, see <http://www.gnu.org/licenses/>.
*/

import { IWaveParameters, IWave, IWaveChannel } from '../node_modules/rvl-node-types/declarations/index';

const EMPTY_CHANNEL: IWaveChannel = { a: 0, w_t: 0, w_x: 0, phi: 0, b: 0 };

export interface IPixel {
  r: number;
  g: number;
  b: number;
}

interface IColor {
  r: number;
  g: number;
  b: number;
  a: number;
}

let nWaves = 0;
let nPixels = 0;
let startTime = 0;
let time = 0;
let wParameters: IWaveParameters;

export function initRenderer(waveParameters: IWaveParameters, numPixels: number, numWaves = 4): void {
  nWaves = numWaves;
  nPixels = numPixels;
  wParameters = waveParameters;
  startTime = Date.now();
}

export function resetRendererClock(): void {
  startTime = Date.now();
}

export function getRendererClock(): number {
  return time;
}

// Modified from https://axonflux.com/handy-rgb-to-hsl-and-rgb-to-hsv-color-model-c
function hsvToRgb(color: number[]): number[] {
  const [ h, s, v] = color;
  let r = 0;
  let g = 0;
  let b = 0;

  const i = Math.floor(h * 6);
  const f = h * 6 - i;
  const p = v * (1 - s);
  const q = v * (1 - f * s);
  const t = v * (1 - (1 - f) * s);

  switch (i % 6) {
    case 0: r = v, g = t, b = p; break;
    case 1: r = q, g = v, b = p; break;
    case 2: r = p, g = v, b = t; break;
    case 3: r = p, g = q, b = v; break;
    case 4: r = t, g = p, b = v; break;
    case 5: r = v, g = p, b = q; break;
  }

  return [ r * 255, g * 255, b * 255 ];
}

function calculatePixelValue(waveChannel: IWaveChannel, t: number, x: number): number {
  // Approximate 8-bit sin function from C++
  let sin = 2 * Math.PI * (waveChannel.w_t * t / 100 + waveChannel.w_x * x + waveChannel.phi) / 255;
  sin = 127.5 * (Math.sin(sin) + 1);
  sin = Math.round(sin * waveChannel.a / 255 + waveChannel.b);
  return sin;
}

export function renderPixels(): IPixel[] {
  if (!calculatePixelValue) {
    throw new Error('renderPixels called before init');
  }
  const animationClock = Date.now() - startTime;
  if (!wParameters.timePeriod) {
    wParameters.timePeriod = 255;
  }
  if (!wParameters.distancePeriod) {
    wParameters.distancePeriod = 32;
  }
  const colors: IColor[][] = [];
  time = animationClock % 25500;
  for (let i = 0; i < nPixels; i++) {
    const pixelColorLayers: IColor[] = [];

    for (let j = 0; j < nWaves; j++) {
      const x = Math.floor(255 * (i % wParameters.distancePeriod) / wParameters.distancePeriod);
      const wave = wParameters.waves[j];
      const pixelColor = hsvToRgb([
        calculatePixelValue(wave.h, time, x) / 255,
        calculatePixelValue(wave.s, time, x) / 255,
        calculatePixelValue(wave.v, time, x) / 255
      ]);
      pixelColorLayers[j] = {
        r: pixelColor[0],
        g: pixelColor[1],
        b: pixelColor[2],
        a: calculatePixelValue(wave.a, time, x)
      };
    }
    colors[i] = pixelColorLayers;
  }
  return colors.map((layer) => {
    const pixel: IPixel = {
      r: layer[nWaves - 1].r,
      g: layer[nWaves - 1].g,
      b: layer[nWaves - 1].b
    };
    function blend(bottom: number, top: number, alpha: number): number {
      alpha = alpha / 255;
      bottom = bottom / 255;
      top = alpha * top / 255;
      return Math.round(255 * (top + bottom * (1 - alpha)));
    }
    for (let i = nWaves - 2; i >= 0; i--) {
      pixel.r = blend(pixel.r, layer[i].r, layer[i].a);
      pixel.g = blend(pixel.g, layer[i].g, layer[i].a);
      pixel.b = blend(pixel.b, layer[i].b, layer[i].a);
    }
    return pixel;
  });
}

export function createEmptyWave(): IWave {
  return {
    h: { ...EMPTY_CHANNEL },
    s: { ...EMPTY_CHANNEL },
    v: { ...EMPTY_CHANNEL },
    a: { ...EMPTY_CHANNEL }
  };
}

export function createWaveParameters(wave1?: IWave, wave2?: IWave, wave3?: IWave, wave4?: IWave): IWaveParameters {
  return {
    waves: [
      wave1 || createEmptyWave(),
      wave2 || createEmptyWave(),
      wave3 || createEmptyWave(),
      wave4 || createEmptyWave()
    ]
  };
}

function validateNum(num: number, min: number, max: number, name: string): void {
  if (typeof num !== 'number' || num < min || num > max) {
    throw new Error(`Invalid ${name} ${num}. ` +
      `${name[0].toUpperCase() + name.substring(1)} must be a number between ${min} and ${max}`);
  }
}

export function createSolidColorWave(h: number, s: number, a: number): IWave {
  validateNum(h, 0, 255, 'hue');
  validateNum(s, 0, 255, 'saturation');
  validateNum(a, 0, 255, 'alpha');
  const wave = createEmptyWave();
  wave.h.b = Math.round(h);
  wave.s.b = Math.round(s);
  wave.v.b = 255;
  wave.a.b = Math.round(a);
  return wave;
}

export function createColorCycleWave(rate: number, a: number): IWave {
  validateNum(rate, 1, 32, 'rate');
  validateNum(a, 0, 255, 'alpha');
  const wave = createEmptyWave();
  wave.h.a = 255;
  wave.h.w_t = Math.round(rate);
  wave.h.w_x = 0;
  wave.s.b = 255;
  wave.v.b = 255;
  wave.a.b = Math.round(a);
  return wave;
}

export function createMovingWave(h: number, s: number, rate: number, spacing: number): IWave {
  validateNum(rate, 0, 32, 'rate');
  validateNum(spacing, 1, 16, 'spacing');
  validateNum(h, 0, 255, 'hue');
  validateNum(s, 0, 255, 'saturation');
  const wave = createEmptyWave();
  wave.h.b = Math.round(h);
  wave.s.b = Math.round(s);
  wave.v.b = 255;
  wave.a.a = 255;
  wave.a.w_t = Math.round(rate);
  wave.a.w_x = Math.round(spacing);
  return wave;
}

export function createPulsingWave(h: number, s: number, rate: number): IWave {
  validateNum(rate, 1, 32, 'rate');
  validateNum(h, 0, 255, 'hue');
  validateNum(s, 0, 255, 'saturation');
  const wave = createEmptyWave();
  wave.h.b = Math.round(h);
  wave.s.b = Math.round(s);
  wave.v.b = 255;
  wave.a.w_t = Math.round(rate);
  wave.a.a = 255;
  return wave;
}

export function createRainbowWave(a: number, rate: number): IWave {
  validateNum(rate, 1, 32, 'rate');
  validateNum(a, 0, 255, 'alpha');
  const wave = createEmptyWave();

  wave.h.a = 255;
  wave.h.w_t = Math.round(rate);
  wave.h.w_x = 2;
  wave.s.b = 255;
  wave.v.b = 255;
  wave.a.b = Math.round(a);
  return wave;
}
